### PR TITLE
neovim, neovimUtils, neovim-qt: drop python2 support

### LINKED ISSUE #121438: neovim fails to build with pynvim-0.4.3
## Issue description

Neovim fails to build with the error:

```
error: pynvim-0.4.3 not supported for interpreter python2.7
```

### Steps to reproduce

Run `nix build ./#neovim` on master.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.2, NixOS, 21.05.20210426.801720a (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210326_dd77f71`
 - channels(root): `"nixos-21.03pre265961.891f607d530"`
 - channels(dbuckley): `""`
 - nixpkgs: `/nix/store/9fj0vhs7cwi7aam911vdrsjccmxd0b2m-source`

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

In 2a00e53bd pynvim support for python2 was disabled, this broke the
neovim build. I really think it is time to let go of python2 support in
neovim.

Fixes the build for me, I have no regrets.

I'm aware we might need a release notes entry for that. I'm open to suggestions, I've never relied on python2 in neovim.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  ```
  /nix/store/synllhwl09qs1jzg9v452npvbgq6y0pf-neovim-0.4.4	  425376328
  /nix/store/zapccq0qrj4k5b94j49ifnqy6i5mzxkx-neovim-0.4.4	  383572776
  ```
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/neovim/qt.nix [hunk 1]
```diff
@@ -1,5 +1,4 @@
-{ lib, stdenv, mkDerivation, fetchFromGitHub, cmake, doxygen, makeWrapper
-, msgpack, neovim, pythonPackages, qtbase, neovim-qt-unwrapped }:
+{ stdenv, makeWrapper, neovim, neovim-qt-unwrapped }:
```
Comment (mweinelt) [inline pkgs/applications/editors/neovim/qt.nix:1]: Drive-by cleanup of inputs, not used since 3a25004f2b33b64b9292d923b1a6257b517157a7.

**Comment by teto** [comment]
it's a bit annoying to break compatibility inside  the same neovim release (python2 is dropped in 0.5) but I am not aware of any python2 remote plugins so it should probably be fine ?

**Comment by teto** [comment]
maybe add some warning if python2 is used since it's a breaking change for the same version so it can be surprising.

**Comment by mweinelt** [comment]
> maybe add some warning if python2 is used since it's a breaking change for the same version so it can be surprising.

Sounds great, though not sure where would be a good spot.

**Comment by teto** [comment]
when `withPython2` is used, add a `lib.warn` or even a `throw` with the appropriate message.

**Comment by mweinelt** [comment]
I've added three asserts/throws, can you check whether that is sufficient?

Also extended the release notes for vims python2 support drop with neovim.

**Review by teto — APPROVED** [review]
Looks good to me

**Comment by austinbutler** [comment]
Hopefully this is just me doing it wrong, but I can't seem to get this to work in my home-manager config.

```nix
{
  programs.neovim = {
    enable = true;
    package = (import (builtins.fetchTarball {
      url =
        "https://github.com/mweinelt/nixpkgs/archive/b91ddf1960e1c825a9bcddcbfa67d4eae0e1b971.tar.gz";
      sha256 = "0jdw4j0r6fkqhqaac5gr4i5nq35pm49wh98idzckrzhsg5h1b7yn";
    }) { config = { allowUnfree = true; }; }).neovim-unwrapped;
    extraConfig = ''
      if (has("nvim"))
        "For Neovim 0.1.3 and 0.1.4 < https://github.com/neovim/neovim/pull/2198 >
        let $NVIM_TUI_ENABLE_TRUE_COLOR=1
      endif

      "For Neovim > 0.1.5 and Vim > patch 7.4.1799 < https://github.com/vim/vim/commit/61be73bb0f965a895bfb064ea3e55476ac175162 >
      "Based on Vim patch 7.4.1770 (`guicolors` option) < https://github.com/vim/vim/commit/8a633e3427b47286869aa4b96f2bfc1fe65b25cd >
      " < https://github.com/neovim/neovim/wiki/Following-HEAD#20160511 >
      if (has("termguicolors"))
        set termguicolors
      endif

      filetype plugin indent on
      syntax on
      set tabstop=4 shiftwidth=4 expandtab
      set clipboard+=unnamedplus
    '';
    extraPython3Packages = (ps: with ps; [ black isort pylint ]);
    plugins = with pkgs.vimPlugins; [
      vim-nix
      coc-nvim
      coc-css
      coc-explorer
      coc-git
      coc-go
      coc-html
      coc-json
      coc-prettier
      coc-pyright
      coc-rust-analyzer
      coc-tsserver
      coc-yaml
      {
        plugin = vim-autoformat;
        # https://github.com/Chiel92/vim-autoformat/blob/master/plugin/defaults.vim
        config = ''
          let g:formatters_javascript = ['prettier']
          let g:formatters_json = ['prettier']
          let g:formatters_python = ['black']
          let g:formatters_typescript = ['prettier']
          let g:formatters_yaml = ['prettier']
        '';
      }
      {
        plugin = vim-one;
        config = ''
          set background=dark
          colorscheme one
        '';
      }
      vim-lastplace
    ];
    viAlias = true;
    vimAlias = true;
    vimdiffAlias = true;
    withNodeJs = true;
    withPython3 = true;
  };
}
```

I still end up with:
```text
unpacking channels...
error: pynvim-0.4.3 not supported for interpreter python2.7
(use '--show-trace' to show detailed location information)
```

It seems it should be using the package from your branch since it checks the sha256.

**Comment by mweinelt** [comment]
The only pynvim occurrence that are left are scoped with `python3.pkgs` or `python3Packages`, so I guess you are not using the matching utils or wrapper from this branch.

**Comment by teto** [comment]
@SCOTT-HAMILTON does this PR work for you ?

**Comment by austinbutler** [comment]
Was able to get it working once I realized `home-manager` calls `wrapNeovimUnstable`, I then created an overlay for that package instead. Sorry for the noise, but I can confirm this works for me now.
